File: pe-06/practice.ts
type Lesson = {
    name: string;
    completed: boolean;
    minutes: number;
};

const lessons: Lesson[] = [
    { name: "Variables", completed: true, minutes: 30 },
    { name: "Loops", completed: false, minutes: 45 },
    { name: "Functions", completed: true, minutes: 40 },
    { name: "Arrays", completed: false, minutes: 25 }
];

function countCompleted(lessons: Lesson[]): number {
    let count = 0;

    for (const lesson of lessons) {
        if (lesson.completed === true) {
            count++;
        }
    }

    return count;
}

function getIncompleteLessons(lessons: Lesson[]): string[] {
    const incomplete: string[] = [];

    for (const lesson of lessons) {
        if (lesson.completed === false) {
            incomplete.push(lesson.name);
        }
    }

    return incomplete;
}

function calculateTotalMinutes(lessons: Lesson[]): number {
    let total = 0;

    for (const lesson of lessons) {
        total += lesson.minutes;
    }

    return total;
}

function printLessons(lessons: Lesson[]): void {
    if (lessons.length === 0) {
        console.log("No lessons.");
        return;
    }

    for (let i = 0; i < lessons.length; i++) {
        console.log(`${i + 1}. ${lessons[i].name}`);
    }
}

function weeklyReport(lessons: Lesson[]): void {
    const completed = countCompleted(lessons);
    const incomplete = getIncompleteLessons(lessons);
    const totalMinutes = calculateTotalMinutes(lessons);

    console.log("\nWeekly Learning Report");
    console.log("----------------------");
    console.log(`Completed lessons: ${completed}`);

    if (incomplete.length === 0) {
        console.log("Incomplete lessons: none");
    } else {
        console.log("Incomplete lessons:");

        for (const name of incomplete) {
            console.log(`- ${name}`);
        }
    }

    console.log(`Total minutes: ${totalMinutes}`);
    console.log("Lessons:");

    printLessons(lessons);
}

function validateInput(input: string): number | null {
    let attempts = 0;

    while (attempts < 3) {
        const number = Number(input);

        if (
            input.trim() !== "" &&
            Number.isInteger(number) &&
            number >= 0
        ) {
            console.log("Valid input:", number);
            return number;
        }

        attempts++;

        console.log(`Invalid input. Attempt ${attempts} of 3.`);

        if (attempts === 3) {
            console.log("Maximum attempts reached.");
            return null;
        }
    }

    return null;
}

weeklyReport(lessons);

console.log("\nInput validation:");
validateInput("abc");

File: pe-06/practice.py
lessons = [
    {"name": "Variables", "completed": True, "minutes": 30},
    {"name": "Loops", "completed": False, "minutes": 45},
    {"name": "Functions", "completed": True, "minutes": 40},
    {"name": "Arrays", "completed": False, "minutes": 25}
]


def count_completed(lessons):
    count = 0

    for lesson in lessons:
        if lesson["completed"] is True:
            count += 1

    return count


def get_incomplete_lessons(lessons):
    incomplete = []

    for lesson in lessons:
        if lesson["completed"] is False:
            incomplete.append(lesson["name"])

    return incomplete


def calculate_total_minutes(lessons):
    total = 0

    for lesson in lessons:
        total += lesson["minutes"]

    return total


def print_lessons(lessons):
    if len(lessons) == 0:
        print("No lessons.")
        return

    for i in range(len(lessons)):
        print(f"{i + 1}. {lessons[i]['name']}")


def weekly_report(lessons):
    completed = count_completed(lessons)
    incomplete = get_incomplete_lessons(lessons)
    total_minutes = calculate_total_minutes(lessons)

    print("\nWeekly Learning Report")
    print("----------------------")
    print(f"Completed lessons: {completed}")

    if len(incomplete) == 0:
        print("Incomplete lessons: none")
    else:
        print("Incomplete lessons:")

        for name in incomplete:
            print(f"- {name}")

    print(f"Total minutes: {total_minutes}")
    print("Lessons:")

    print_lessons(lessons)


def validate_input(input_value):
    attempts = 0

    while attempts < 3:
        try:
            number = int(input_value)

            if number >= 0:
                print("Valid input:", number)
                return number

        except ValueError:
            pass

        attempts += 1
        print(f"Invalid input. Attempt {attempts} of 3.")

        if attempts == 3:
            print("Maximum attempts reached.")
            return None

    return None


weekly_report(lessons)

print("\nInput validation:")
validate_input("abc")